### PR TITLE
feat(vlm): add Qwen3-VL support and cookbook

### PR DESCRIPTION
**Pull request by Yanko-7**
## Motivation

This PR adds end-to-end serving support for `Qwen/Qwen3-VL-32B-Instruct` and documents the validated deployment path.

The existing Qwen3 implementation supports text-only inference, but it does not provide the Qwen3-VL vision encoder, multimodal position handling, visual embedding merge, or deepstack integration required to serve Qwen3-VL checkpoints. This change adds those model components and connects them to the existing in-model multimodal runtime while preserving the text-only Qwen3 path.

Additionally, this PR adds a Qwen3-VL cookbook covering deployment, OpenAI-compatible text/image/video invocation, MMMU and MMMU-Pro Vision accuracy, and a single-image multimodal serving benchmark.

### Qwen3-VL model support

- Register `Qwen3VLForConditionalGeneration` in the model registry.
- Add the Qwen3-VL vision encoder, patch embedding, rotary position embedding, transformer blocks, merger, and multimodal projector.
- Reuse the Qwen3 language-model implementation for Qwen3-VL while adding multimodal RoPE, visual embedding injection, and deepstack feature support.
- Preserve the existing text-only Qwen3 execution path.

### Multimodal runtime integration

- Extend compilation management to precompile the required multimodal execution shapes and vision patch buckets.
- Unify text and multimodal EXTEND signatures for in-model VLMs, reducing model-forward EXTEND precompile variants from two to one per token bucket.
- Update model-runner and host-orchestration paths to carry and apply Qwen3-VL deepstack embeddings.
- Keep text-only and multimodal execution stable through the shared in-model multimodal path.
- Set BF16 as the default dtype for the shared multimodal lane-packing helper.
- Add regression coverage for multimodal compilation and in-model multimodal serving.

### Documentation

- Add the Qwen3-VL cookbook and register it in the cookbook navigation.
- Document the validated TPU topology, launch flags, multimodal worker configuration, invocation examples, accuracy results, and single-image multimodal performance results.

## Accuracy Tests

| Item | Value |
| --- | --- |
| Model | `Qwen/Qwen3-VL-32B-Instruct` |
| Hardware | Single-host TPU v7x-8, topology `2x2x1` |
| Parallelism | DP4 x effective TP2 (`--tp-size 8 --dp-size 4`) |
| Precision | BF16 model, vision encoder, and KV cache |
| Context / max sequence | 32768 / 32768 |
| Evaluator | EvalScope 1.5.1, OpenAI-compatible API |
| Generation | `max_tokens=8192`, temperature 0, seed 42 |
| Evaluation batch size | 24 for MMMU; 256 for MMMU-Pro Vision |

### Server command

```bash
export MODEL_PATH=Qwen/Qwen3-VL-32B-Instruct
export JAX_COMPILATION_CACHE_DIR=/tmp/jit_cache

python -u -m sgl_jax.launch_server \
  --model-path "$MODEL_PATH" \
  --trust-remote-code \
  --skip-server-warmup \
  --device tpu \
  --tp-size 8 \
  --dp-size 4 \
  --dtype bfloat16 \
  --kv-cache-dtype bf16 \
  --context-length 32768 \
  --max-seq-len 32768 \
  --max-running-requests 1024 \
  --max-prefill-tokens 16384 \
  --chunked-prefill-size 4096 \
  --mem-fraction-static 0.8 \
  --page-size 128 \
  --disable-radix-cache \
  --vision-encoder-parallel dp \
  --mm-io-worker-num 4 \
  --mm-processor-worker-num 2 \
  --random-seed 0 \
  --host 0.0.0.0 \
  --port 30000
```

### Evaluation command

Install the evaluator and save the following script as `eval_accuracy.py`:

```bash
pip install 'evalscope[app,perf]==1.5.1'
```

```python
import os

from evalscope import TaskConfig, run_task

dataset_args = {
    "mmmu": {"dataset_id": "MMMU/MMMU"},
    "mmmu_pro": {
        "dataset_id": "MMMU/MMMU_Pro",
        "extra_params": {"dataset_format": "vision"},
    },
}

for dataset, batch_size in (("mmmu", 24), ("mmmu_pro", 256)):
    task_cfg = TaskConfig(
        model="Qwen/Qwen3-VL-32B-Instruct",
        api_url="http://127.0.0.1:30000/v1",
        api_key="EMPTY",
        eval_type="openai_api",
        datasets=[dataset],
        dataset_hub="huggingface",
        dataset_args={dataset: dataset_args[dataset]},
        eval_batch_size=batch_size,
        generation_config={
            "max_tokens": 8192,
            "temperature": 0.0,
            "stream": False,
        },
        seed=42,
        work_dir=os.path.join("./outputs/qwen3vl32b_vlm", dataset),
    )
    run_task(task_cfg=task_cfg)
```

```bash
python eval_accuracy.py
```

### Results

| Dataset | Metric | Result | Samples |
| --- | --- | ---: | ---: |
| MMMU Val | Mean accuracy | **72.22%** | 900 |
| MMMU-Pro Vision | Mean accuracy | **62.43%** | 1,730 |

### Precompile reduction

In-model VLMs previously compiled separate text and multimodal model-forward EXTEND variants for every token bucket. Text requests now use the same multimodal-shaped input signature, with zero-filled deepstack tensors and a runtime `apply_for_deepstack` flag when no visual input is present.

This reduces model-forward EXTEND precompile invocations from two to one per token bucket, a **50% reduction**. Vision encoder and embedding-merge bucket precompilation are unchanged, so this is not a claim that total server startup time is reduced by 50%.

### Setup

Both runs use `Qwen/Qwen3-VL-32B-Instruct` on a single TPU v7x-8 (`2x2x1`), DP4 x effective TP2, BF16 model/vision/KV, no quantization, 1,000 requests at an unbounded request rate, one warmup request, a cache flush, 1,024 random source-text tokens, and 500 output tokens per request.

The text run uses a 32,768-token context. The multimodal run uses a 2,048-token context and additionally sends one random 512 x 512 JPEG per request with 4 I/O workers and 2 processor workers. After chat templating and image processing, each multimodal request averages 1,086.25 text tokens and 258 vision tokens, or 1,344.25 total input tokens.

These are two independently measured serving workloads, not an equal-input A/B comparison.

### Text server command

The text benchmark reuses the 32K server command from the accuracy section.

### Text benchmark command

```bash
python -m sgl_jax.bench_serving \
  --backend sglang-oai-chat \
  --host 127.0.0.1 --port 30000 \
  --model "$MODEL_PATH" --tokenizer "$MODEL_PATH" \
  --dataset-name random \
  --num-prompts 1000 \
  --random-input-len 1024 \
  --random-output-len 500 \
  --random-range-ratio 1.0 \
  --request-rate inf \
  --seed 0 \
  --warmup-requests 1 \
  --flush-cache --output-details
```

### Multimodal server command

```bash
export MODEL_PATH=Qwen/Qwen3-VL-32B-Instruct
export JAX_COMPILATION_CACHE_DIR=/tmp/jit_cache

python -u -m sgl_jax.launch_server \
  --model-path "$MODEL_PATH" \
  --trust-remote-code \
  --skip-server-warmup \
  --device tpu \
  --tp-size 8 \
  --dp-size 4 \
  --dtype bfloat16 \
  --kv-cache-dtype bf16 \
  --context-length 2048 \
  --max-seq-len 2048 \
  --max-running-requests 1024 \
  --max-prefill-tokens 16384 \
  --chunked-prefill-size 4096 \
  --mem-fraction-static 0.9 \
  --page-size 128 \
  --disable-radix-cache \
  --vision-encoder-parallel dp \
  --mm-io-worker-num 4 \
  --mm-processor-worker-num 2 \
  --random-seed 0 \
  --host 0.0.0.0 \
  --port 30000
```

### Multimodal benchmark command

```bash
python -m sgl_jax.bench_serving \
  --backend sglang-oai-chat \
  --host 127.0.0.1 --port 30000 \
  --model "$MODEL_PATH" --tokenizer "$MODEL_PATH" \
  --dataset-name image \
  --num-prompts 1000 \
  --random-input-len 1024 \
  --random-output-len 500 \
  --random-range-ratio 1.0 \
  --image-count 1 \
  --image-resolution 512x512 \
  --image-format jpeg \
  --image-content random \
  --request-rate inf \
  --seed 0 \
  --warmup-requests 1 \
  --flush-cache --output-details
```

### Results

| Metric | Text | Single-image multimodal |
| --- | ---: | ---: |
| Successful requests | 1,000 | 1,000 |
| Duration | 58.720 s | 64.332 s |
| Request throughput | 17.030 req/s | 15.544 req/s |
| Input token throughput | 17,438.83 tok/s | 20,895.61 tok/s |
| Output token throughput | **8,515.05 tok/s** | **7,772.22 tok/s** |
| Total token throughput | 25,953.88 tok/s | 28,667.82 tok/s |
| Mean TTFT | 14,085.46 ms | 16,636.34 ms |
| Median TTFT | 9,925.04 ms | 13,818.60 ms |
| P99 TTFT | 47,376.80 ms | 54,075.23 ms |
| Mean TPOT | 65.75 ms | 73.94 ms |
| Median TPOT | 70.23 ms | 77.37 ms |
| P99 TPOT | 84.61 ms | 100.26 ms |
| Median E2E latency | 44,980.55 ms | 52,606.81 ms |
| P99 E2E latency | 58,337.06 ms | 63,657.17 ms |

## Test Plan

Targeted regression tests:

TPU serving validation:

- [x] 1,000-request text benchmark.
- [x] 1,000-request single-image multimodal benchmark.
- [x] MMMU Val evaluation with all 900 samples completed.
- [x] MMMU-Pro Vision evaluation with all 1,730 samples completed.
- [x] Unified text/multimodal EXTEND precompile signature regression coverage.
- [x] Cookbook JSON, internal links, Markdown code fences, and formatting checks.

## Checklist

- [x] Please use English, otherwise it will be closed.
- [x] The purpose of the PR, or link existing issues this PR will resolve.
- [x] The test plan, such as providing test command.
- [x] (Optional) The necessary documentation update.

### CODE REVIEW

**Comment by cjx0709** [comment]
@Yanko-7 we just merged some cookbook change, you can solve the conflicts and follow the newest pr in this pr.

File: python/sgl_jax/srt/multimodal/in_model/lane_packing.py [hunk 1]
```diff
@@ -314,7 +314,7 @@ def precompile_mrope_vision_model(
     patch_dim: int,
     merge_unit: int,
     rope_type: Literal["rope_3d", "rope_2d", "rope_2d_packed"],
-    dtype: np.dtype | type,
+    dtype: np.dtype | type = jnp.bfloat16,
```
Comment (lianfang0320) [inline python/sgl_jax/srt/multimodal/in_model/lane_packing.py:317]: Why dtype is hard code as `jnp.bfloat16`? 
Could we pass dtype=self.dtype here, as Qwen2.5-VL does? Otherwise, non-BF16 --dtype settings still pack vision inputs as BF16. If Qwen3-VL is intentionally BF16-only, should we validate that explicitly?

Comment (Yanko-7) [reply python/sgl_jax/srt/multimodal/in_model/lane_packing.py:317]: qwen3VL omitted the dtype; it has now been added.

**Review by JamesBrianD — APPROVED** [review]
